File: src/main/java/com/hackclub/hccore/listeners/AFKListener.java
package com.hackclub.hccore.listeners;

import com.hackclub.hccore.HCCorePlugin;
import com.hackclub.hccore.PlayerData;
import com.hackclub.hccore.events.player.PlayerAFKStatusChangeEvent;
import com.hackclub.hccore.playermessages.afk.IsAFKMessage;
import com.hackclub.hccore.playermessages.afk.NotAFKMessage;
import com.hackclub.hccore.playermessages.afk.RunAFKMessage;
import com.hackclub.hccore.playermessages.afk.YouAFKMessage;
import io.papermc.paper.event.player.AsyncChatEvent;
import java.time.Duration;
import java.time.temporal.ChronoUnit;
import net.kyori.adventure.title.Title;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerMoveEvent;

public class AFKListener implements Listener {

  private final Title afkTitle;

  private final HCCorePlugin plugin;

  public AFKListener(HCCorePlugin plugin) {
    this.plugin = plugin;

    this.afkTitle = Title.title(YouAFKMessage.get(),
        RunAFKMessage.get(), Title.Times.times(
            Duration.of(500, ChronoUnit.MILLIS), Duration.of(12, ChronoUnit.HOURS),
            Duration.of(1, ChronoUnit.SECONDS)));
  }

  @EventHandler
  public void onPlayerAFKStatusChange(final PlayerAFKStatusChangeEvent event) {
    Player player = event.getPlayer();
    PlayerData data = this.plugin.getDataManager().getData(player);

    if (event.getNewValue()) {
      player.showTitle(afkTitle);
      player.getServer().broadcast(IsAFKMessage.get(data.getUsableName(), data.getNameColor()));
    } else {
      player.clearTitle();
      player.getServer().broadcast(NotAFKMessage.get(data.getUsableName(), data.getNameColor()));
    }
  }

  @EventHandler
  public void onAsyncPlayerChat(final AsyncChatEvent event) {
    this.plugin.getDataManager().getData(event.getPlayer())
        .setLastActiveAt(System.currentTimeMillis());
  }

  @EventHandler
  public void onPlayerCommandPreprocess(final PlayerCommandPreprocessEvent event) {
    this.plugin.getDataManager().getData(event.getPlayer())
        .setLastActiveAt(System.currentTimeMillis());
  }

  @EventHandler
  public void onPlayerInteract(final PlayerInteractEvent event) {
    this.plugin.getDataManager().getData(event.getPlayer())
        .setLastActiveAt(System.currentTimeMillis());
  }

  @EventHandler
  public void onPlayerMove(final PlayerMoveEvent event) {
    Location from = event.getFrom();
    Location to = event.getTo();

    // Make sure the player didn't just turn their head
    if (from.toVector().equals(to.toVector())) { // NOTE: Might be inefficient?
      return;
    }

    this.plugin.getDataManager().getData(event.getPlayer())
        .setLastActiveAt(System.currentTimeMillis());
  }
}

File: src/main/java/com/hackclub/hccore/listeners/AFKListener.kt
package com.hackclub.hccore.listenersimport

import org.bukkit.ChatColor
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerCommandPreprocessEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.player.PlayerMoveEvent

com.hackclub.hccore.HCCorePlugin
import com.hackclub.hccore.PlayerData
import com.hackclub.hccore.tasks.CheckAdAstraTask
import com.hackclub.hccore.utils.gson.GsonUtil
import com.hackclub.hccore.utils.gson.LocationSerializer
import com.hackclub.hccore.utils.gson.LocationMapDeserializer
import com.hackclub.hccore.utils.gson.LocationMapSerializer
import com.hackclub.hccore.utils.gson.LocationDeserializer
import java.util.LinkedHashMap
import com.hackclub.hccore.events.player.PlayerAFKStatusChangeEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.events.ReadyEvent
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.api.JDABuilder
import javax.security.auth.login.LoginException
import java.lang.Exception
import com.hackclub.hccore.commands.ColorCommand
import java.util.stream.Collectors
import com.hackclub.hccore.commands.StatsCommand
import java.text.SimpleDateFormat
import com.hackclub.hccore.commands.UpvoteCommand
import com.hackclub.hccore.commands.DownvoteCommand
import com.hackclub.hccore.commands.TableflipCommand
import java.lang.Runnable
import com.hackclub.hccore.listeners.SleepListener
import com.comphenix.protocol.events.ListenerPriority
import com.comphenix.protocol.PacketType
import com.comphenix.protocol.events.PacketAdapter
import com.comphenix.protocol.events.PacketEvent
import com.comphenix.protocol.wrappers.EnumWrappers
import com.comphenix.protocol.wrappers.WrappedGameProfile
import org.bukkit.craftbukkit.v1_17_R1.advancement.CraftAdvancement
import net.minecraft.advancements.AdvancementDisplay
import net.minecraft.network.chat.IChatBaseComponent
import net.minecraft.advancements.AdvancementFrameType
import de.tr7zw.changeme.nbtapi.NBTTileEntity
import java.io.FileReader
import java.io.IOException
import java.io.FileWriter
import java.util.UUID
import java.util.HashMap
import com.hackclub.hccore.DataManager
import com.comphenix.protocol.ProtocolManager
import com.hackclub.hccore.discord.DiscordBot
import com.comphenix.protocol.ProtocolLibrary
import com.hackclub.hccore.commands.AFKCommand
import com.hackclub.hccore.commands.LocCommand
import com.hackclub.hccore.commands.NickCommand
import com.hackclub.hccore.commands.PingCommand
import com.hackclub.hccore.commands.ShrugCommand
import com.hackclub.hccore.commands.SpawnCommand
import com.hackclub.hccore.commands.DiscordCommand
import com.hackclub.hccore.listeners.AdvancementListener
import com.hackclub.hccore.listeners.AFKListener
import com.hackclub.hccore.listeners.BeehiveInteractionListener
import com.hackclub.hccore.listeners.PlayerListener
import com.hackclub.hccore.listeners.NameChangeListener
import com.hackclub.hccore.tasks.AutoAFKTask
import hu.trigary.advancementcreator.AdvancementFactory
import hu.trigary.advancementcreator.shared.ItemObject

class AFKListener(plugin: HCCorePlugin) : Listener {
    private val plugin: HCCorePlugin
    @EventHandler
    fun onPlayerAFKStatusChange(event: PlayerAFKStatusChangeEvent) {
        val player = event.player
        val data: PlayerData = plugin.getDataManager().getData(player)
        if (event.newValue) {
            player.sendTitle(ChatColor.RED.toString() + ChatColor.BOLD.toString() + "You are AFK",
                    "Run /afk to mark yourself as active", 10, 999999, 20)
            player.server.broadcastMessage(data.usableName + " is now AFK")
        } else {
            player.sendTitle(null, null, 0, 1, -1)
            player.server.broadcastMessage(data.usableName + " is now active")
        }
    }

    @EventHandler
    fun onAsyncPlayerChat(event: AsyncPlayerChatEvent) {
        plugin.getDataManager().getData(event.player)
                .setLastActiveAt(System.currentTimeMillis())
    }

    @EventHandler
    fun onPlayerCommandPreprocess(event: PlayerCommandPreprocessEvent) {
        plugin.getDataManager().getData(event.player)
                .setLastActiveAt(System.currentTimeMillis())
    }

    @EventHandler
    fun onPlayerInteract(event: PlayerInteractEvent) {
        plugin.getDataManager().getData(event.player)
                .setLastActiveAt(System.currentTimeMillis())
    }

    @EventHandler
    fun onPlayerMove(event: PlayerMoveEvent) {
        val from = event.from
        val to = event.to

        // Make sure the player didn't just turn their head
        if (from.toVector() == to!!.toVector()) { // NOTE: Might be inefficient?
            return
        }
        plugin.getDataManager().getData(event.player)
                .setLastActiveAt(System.currentTimeMillis())
    }

    init {
        this.plugin = plugin
    }
}